state/application.go

```diff
@@ -3124,6 +3132,83 @@ func allUnits(st *State, application string) (units []*Unit, err error) {
 	return units, nil
 }
 
+// GetUnitAttachmentInfos returns a list of UnitAttachmentInfo for this application.
+func (a *Application) GetUnitAttachmentInfos(statuses []status.Status) (unitAttachmentInfos []UnitAttachmentInfo, err error) {
+	logger.Warningf("jneo8 unitCount: %d scale: %d", a.doc.UnitCount, a.doc.DesiredScale)
+	// In case of scale down, there won't have attachment info.
+	if a.doc.DesiredScale < a.doc.UnitCount {
+		return unitAttachmentInfos, nil
+	}
+
+	unitIds := []string{}
+
+	// When deploying an application, UnitCount will match DesiredScale (which defaults to 1).
+	// Note: Attaching storage is not allowed during deployment if DesiredScale > 1, so that case is ignored.
+	// When using add-unit, DesiredScale will be greater than UnitCount during the initial part of the lifecycle.
+	if (a.doc.UnitCount == 1 && a.doc.DesiredScale == 1) || (a.doc.UnitCount < a.doc.DesiredScale) {
```

Comment:
This can be simplified and moved to the check above so here we'' just do the for loop. Also if desired scale < unit count then the for loop will exit anyway and the len unitIds check will return early.